php-frontend/src/test/java/org/sonar/php/cfg/ControlFlowGraphTest.java

```diff
@@ -46,133 +45,128 @@
  * -- 'pred' is a bracketed array of expected predecessor ids
  *
  * 2. each basic block must contain the 'block' function call as the first statement
- *
- * 3. the metadata string is constructed with the {@link BlockString}
  */
 public class ControlFlowGraphTest extends PHPTreeModelTest {
 
-  /**
-   * The predecessors are constructed based on the successors, so we should not test them all the time
-   */
   @Test
-  public void if_then_test_predecessors() {
-    ControlFlowGraph cfg = build(functionTree(
-      BlockString.withId(0).withSuccessorIds(1, 2).withPredecessorIds().build() +
-        "$x = 1;" +
-        "1 + 1;" +
+  public void test_parsing_for_start_block() {
+    // this is a test for the ExpectedStructureParser
+    ControlFlowGraph cfg = cfgForFunctionBody(
+      "block( [ id => 0, succ => [1,2], pred => [] ] );" +
         "if (a) {" +
-        BlockString.withId(1).withSuccessorIds(2).withPredecessorIds(0).build() +
+        "  block( [ id => 1, succ => [2], pred => [0] ] );" +
         "}" +
-        BlockString.withId(2).withSuccessorIds("END").withPredecessorIds(0, 1).build()));
+        "block( [ id => 2, succ => [END], pred => [0,1] ] );");
 
-    assertThat(cfg.blocks()).hasSize(4);
+    Map<String, ExpectedBlockStructure> expected = ExpectedStructureParser.parse(cfg.blocks());
+    assertThat(cfg.start()).isEqualTo(expected.get("0").actualBlock());
+  }
 
-    Map<String, BlockMetadata> data = MetadataParser.parse(cfg.blocks());
+  @Test
+  public void test_end_predecessors() {
+    ControlFlowGraph cfg = cfgForFunctionBody(
+      "foo();" +
+        "if (a) {" +
+        "  bar();" +
+        "  if (b) {" +
+        "    qix();" +
+        "  }" +
+        "}");
 
-    CfgBlock startBlock = data.get("0").getBlock();
-    assertThat(startBlock).isEqualTo(cfg.start());
-    assertThat(startBlock).isExactlyInstanceOf(PhpCfgBranchingBlock.class);
+    assertThat(cfg.end().predecessors()).hasSize(1);
```

Comment:
I think it should be 3
And we have one because we didn't remove empty blocks. I think it should be done in this PR